# MULTICOM4
This is the development repository for the MULTICOM protein structure prediction system for the CASP16 competition. It includes all the programs used in the CASP15 experiment and new programs developed for the CASP16 experiment. 

[Notes for updated tools](Tools_notes.md)

[Notes for updated databases](Database_notes.md)

Monomer

- Adjustable inputs for AlphaFold2-based variants

     | Adjustable inputs  | Description | Choices | default value |
     | -------------| ---------------| ------ | ------ |
     | MSA inputs| MSA sampling | default, original, colabfold, img, deepmsa, esm_msa | default | 
     | Template inputs | Template sampling | pdb70 (default), pdb_sort90 (inhouse template database), tmsearch, foldseek, no templates | pdb70 (default) |
     | monomer_num_ensemble | number of ensemble | 8 (casp15) |  3 |
     | monomer_num_recycle | number of recycle | 8 (casp15) | 1 |
     | monomer preset | deep learning models for monomer | monomer, monomer_ptm | monomer |
     | dropout | dropout rate during inference | true, false | false |
     | dropout_structure_module | Dropout in structure module at inference | true, false | true (0.1) |
     | predictions_per_model | number of predictions for each model | 1 (casp15) | 1 |
     | relax_topn_predictions | number of predictors that will be relaxed | ALL | ALL |

- AlphaFold2-based predictors

     | Predictor  | MSA | Template | monomer preset | dropout | dropout in structure module | Description |
     | -------- | ----- | ----- | ----- | -----| ----- | ----- |
     | default | default | pdb70 (default) | monomer | false | true (0.1) | default AlphaFold2 | 
     | default_seq_temp | default | pdb_sort90 | monomer | false | true (0.1) | Inhouse template database |
     | original | original | pdb70 (default) | monomer | false | true (0.1) | search MSA seperately |
     | ori_seq_temp | original | pdb_sort90 | monomer | false | true (0.1) | Inhouse template database|
     | colabfold | colabfold | pdb70 (default) | monomer | false | true (0.1) | ColabFold Alignment | 
     | colab_seq_temp | colabfold | pdb_sort90 | monomer | false | true (0.1) | Inhouse template database|
     | img | img | pdb70 (default) | monomer | false | true (0.1) | IMG alignment |
     | img_seq_temp | img | pdb_sort90 | monomer | false | true (0.1) | Inhouse template database|
     | dhr | DHR | pdb70 (default) | monomer | false | true (0.1) | DHR Alignment generated by Pawan's program |
     | def_drop_s | default | pdb70 (default) | monomer | true | true (0.1) | AFSample | 
     | def_drop_nos | default | pdb70 (default) | monomer | true | false | AFSample | 
     | def_notemp | default | no templates | monomer | false | true (0.1) | AFSample |
     | def_notemp_drop_s | default | no templates | monomer | true | true (0.1) | AFSample |
     | def_notemp_drop_nos | default | no templates | monomer | true | false | AFSample |
     | DeepMSA predictors | DeepMSA | pdb70 (default) | monomer | false | true (0.1) | DeepMSA alignments (up to 16) |

     - Post-AlphaFold2-based predictors (requires models generated by default as inputs)

          | Predictor  | MSA | Template | monomer preset | dropout | dropout in structure module | Description |
          | -------- | ----- | ----- | ----- | ----- | ----- | ----- |
          | def_esm_msa | Default + MSAs generated by esm | pdb70 (default) | monomer false | true (0.1) | Use esm to generate MSA |
          | foldseek_refine | Default + Foldseek (pdb + af_db) | Foldseek | monomer | false | true (0.1) | Foldseek refinement using the top-ranked models generated by default alphafold2, number of input models: 5; number of output models: 5 |
          | foldseek_refine_esm | Default + Foldseek (esm_atlas) | pdb70 (default) | monomer | false | true (0.1) | Foldseek refinement using the top-ranked models generated by default alphafold2, number of input models: 5; number of output models: 5 |
          | foldseek_refine_esm_h | Default + Foldseek (esm_atlas, high confidence structures) | pdb70 (default) | monomer | false | true (0.1) | Foldseek refinement using the top-ranked models generated by default alphafold2, number of input models: 5; number of output models: 5 |
          | default_tmsearch | default | tmsearch | monomer | false | true (0.1)  | Use TMsearch to search for pdb templates| 

     Number of models to generate:

     About 100 models per AlphaFold2-based variants (MSA, template, dropout), for non-alphafold predictors, use their default settings to generate models.

     For hard targets, use AFsample to generate 500 models per variants (total 2500) in AFsample. For other predictors, increase to 200 models per AlphaFold2-based variants.

- Non-AlphaFold2-based predictors

| Predictor  | Note |
| -------------| -----| 
| Paddle | Paddle-Helix| 
| ESMFold | ESMFold|
| DeepFold | DeepFold |
| MEGAFold | MEGAFold |

## **Steps**

### Server
     ```
     cd /bmlfast/bml_casp16/MULTICOM4
     source /bmlfast/bml_casp16/anaconda3/bin/activate
     conda activate multicom4 
     export PYTHONPATH=/bmlfast/bml_casp16/MULTICOM4
 
     python python bin/monomer/monomer_alignment.py --option_file bin/db_option_default_local --fasta_path examples/fasta/T1131.fasta --output_dir examples/T1131
     # if running on hellbender, use option_file: db_option_default_hell

     python bin/monomer/monomer_af.py --option_file bin/db_option_default_local --fasta_path examples/fasta/T1131.fasta --output_dir examples/T1131
     # bash scripts will be generated in examples/T1131/N3_monomer_structure_generation/bash_scripts
     # One can run the scripts on different servers

     python bin/monomer/monomer_non_af.py --option_file bin/db_option_default_local --fasta_path examples/fasta/T1131.fasta --output_dir examples/T1131
     # This script will run the non-alphafold2 predictors (don't need to manually run the scripts)

     python bin/monomer/monomer_post_af.py --option_file bin/db_option_default_local --fasta_path examples/fasta/T1131.fasta --output_dir examples/T1131
     # bash scripts will be generated in examples/T1131/N3_monomer_structure_generation/post_def_bash_scripts
     # These predictors take the top-ranked models in default alphafold2 as input
     # Run the scripts after the default predictors have finished

     bin/monomer/monomer_ranking.py --option_file bin/db_option_default_local --fasta_path examples/fasta/T1131.fasta --output_dir examples/T1131
     # Rank the models based on the finished predictors (finished: has ranking_debug.json in the directory)
     ```

### Hellbender
     ```
     cd /cluster/pixstor/chengji-lab/bml_casp16/MULTICOM4
     source /cluster/pixstor/chengji-lab/bml_casp16/anaconda3/bin/activate
     conda activate multicom4 
     export PYTHONPATH=/cluster/pixstor/chengji-lab/bml_casp16/MULTICOM4
     export PERL5LIB=/cluster/pixstor/chengji-lab/bml_casp16/tools/DNCON4/tools/SCRATCH-1D_1.1
     
     # Need at least 300G memory, can only be run on the H100 nodes
     python python bin/monomer/monomer_alignment.py --option_file bin/db_option_default_hell --fasta_path examples/fasta/T1104.fasta --output_dir examples/T1104

     # Run the alphafold2 predictors, the jobs will be automatically submitted
     # Can run multiple times to check whether some of the predictors failed to generate models
     python bin/monomer/monomer_af.py --option_file bin/db_option_default_hell --fasta_path examples/fasta/T1104.fasta --output_dir examples/T1104

     python bin/monomer/monomer_post_af.py --option_file bin/db_option_default_hell --fasta_path examples/fasta/T1104.fasta --output_dir examples/T1104
     # Note: esm_msa needs at least 100G RAM to run
     # Note: programmed to run these predictors on A100 (cannot run esmfold on H100 due to pytorch version issue)

     bin/monomer/monomer_ranking.py --option_file bin/db_option_default_hell --fasta_path examples/fasta/T1131.fasta --output_dir examples/T1131
     # Rank the models based on the finished predictors (finished: has ranking_debug.json in the directory)
     # Note: currently disabled gate ranking (one issue to solve)
     ```

## **Output**

```
$OUTPUT_DIR/                                   # Your output directory
    N1_monomer_alignments_generation/          # Working directory for generating monomer MSAs
        # Note, in addition to the MSAs, disorder predictions, contact map and distance maps are also generated
    N1_monomer_alignments_generation_img/      # Working directory for generating IMG MSA
        # Note: the img.running file may use many disk space
    N2_monomer_template_search/                # Working directory for searching monomer templates
    N3_monomer_structure_generation/           # Working directory for generating monomer structural predictions
        - bash_scripts
          # Bash files to run the predictors
        - post_def_bash_scripts
          # Bash files for the predictors that requires models from default as input
    N4_monomer_structure_evaluation/           # Working directory for evaluating the monomer structural predictions
        - alphafold_ranking.csv    # AlphaFold2 pLDDT ranking
        - pairwise_ranking.tm      # Pairwise (APOLLO) ranking
        - pairwise_af_avg.ranking  # Average ranking of the two
        - gate.csv                 # Gate ranking scores
        - gate/cluster.txt         # Clustered models
        - gate_af_avg.ranking      # Average score of gate and AlphaFold2 pLDDT score
        - ai[1-5].pdb              # Selected models for MULTICOM-AI (AlphaFold pLDDT score, TS similarity TM-score < 0.8)
        - gate[1-5].pdb            # Selected models for MULTICOM-GATE (Gate ranking score, different clusters from gate)
        - llm[1-5].pdb             # Selected models for MULTICOM-LLM (Average score of Gate and AlphaFold pLDDT, different clusters from gate)
```

b. Multimer

- Adjustable inputs for AlphaFold-Multimer-based variants

     | Adjustable inputs  | Description | Choices | default value |
     | -------------| ---------------| ------ | ------ |
     | $MSA_{unpaired}$ inputs| Monomer MSAs | default, deepmsa | default | 
     | $MSA_{paired}$ inputs| Interaction sources | default, species, string, pdb, unidist, deepmsa | default | 
     | Template inputs | Template sampling | pdb_seqres (default), pdb70, pdb, pdb_complex (inhouse template database), structural template (foldseek), no templates | pdb_seqres (default) |
     | multimer_num_ensemble | number of ensemble | 3 (casp15) |  1 |
     | multimer_num_recycle | number of recycle | 5 (casp15) | 20 |
     | dropout | dropout rate during inference | true, false | false |
     | dropout_structure_module | Dropout in structure module at inference | true, false | true (0.1) |
     | predictions_per_model | number of predictions for each model | 5 (casp15) | 5 |
     | relax_topn_predictions | number of predictors that will be relaxed | ALL | ALL |

- AlphaFold-Multimer-based predictors (hetero-multimer)

     | Predictor  | $MSA_{paired}$ | $MSA_{unpaired}$ | Template | dropout | dropout in structure module | Description |
     | -------------| ---------------| ------ | ------ | ---------------| ------ | ------ |
     | default_multimer | default (uniprot) | default | pdb_seqres | false | true (0.1) | default AlphaFold-Multimer |
     | def_mul_nopair | None | default | pdb_seqres | false | true (0.1) | default_multimer no msa pairing | 
     | def_mul_struct | default (uniprot) | default | structural template (foldseek) |  false | true (0.1) | def_mul_struct |
     | def_mul_tmsearch | default (uniprot) | default | structural template (tmsearch) |  false | true (0.1) | def_mul_tmsearch |
     | def_mul_pdb70 | default (uniprot) | default | pdb70 | false | true (0.1) | def_mul_pdb70 |
     | def_mul_pdb | default (uniprot) | default | pdb | false | true (0.1) | def_mul_pdb |
     | def_mul_comp | default (uniprot) | default | pdb_complex | false | true (0.1) | def_mul_comp |
     | uniclust_ox_a3m | species (unclust30) | default | pdb_seqres | false | true (0.1) | uniclust_oxmatch_a3m |
     | pdb_inter_ref_a3m | pdb (uniref30) | default | pdb_seqres | false | true (0.1) | pdb_iter_uniref_a3m |
     | pdb_inter_ref_sto | pdb (uniref90) | default | pdb_seqres | false | true (0.1) | pdb_iter_uniref_sto | 
     | pdb_inter_prot_sto | pdb (uniprot) | default | pdb_seqres | false | true (0.1) | pdb_iter_uniprot_sto |
     | spec_inter_ref_a3m | species (uniref30) | default | pdb_seqres | false | true (0.1) | spec_iter_uniref_a3m |
     | spec_struct | species (uniref30) | default | structural template (foldseek) | false | true (0.1) | spec_struct |
     | spec_pdb70 | species (uniref30) | default | pdb70 | false | true (0.1) | spec_pdb70 |
     | spec_pdb | species (uniref30) | default | pdb | false | true (0.1) | spec_pdb |
     | spec_comp | species (uniref30) | default | pdb_complex | false | true (0.1) | spec_comp |
     | spec_inter_ref_sto | species (uniref90) | default | pdb_seqres | false | true (0.1) | spec_iter_uniref_sto |
     | spec_inter_prot_sto | species (uniprot) | default | pdb_seqres | false | true (0.1) | spec_iter_uniprot_sto |
     | unidist_ref_a3m | unidist (uniref30) | default | pdb_seqres | false | true (0.1) | unidist_uniref_a3m |
     | unidist_ref_sto | unidist (uniref90) | default | pdb_seqres | false | true (0.1) | unidist_uniref_sto |
     | unidist_prot_sto | unidist (uniprot) | default | default | false | true (0.1) | unidist_uniprot_sto |
     | str_inter_ref_a3m | string (uniref30) | default | pdb_seqres | false | true (0.1) | str_iter_uniref_a3m |
     | str_inter_ref_sto | string (uniref90) | default | pdb_seqres | false | true (0.1) | str_iter_uniref_sto |
     | str_struct | string (uniref90) | default | default | structural template (foldseek) | false | true (0.1) | str_struct |
     | str_pdb70 | string (uniref90) | default | default | pdb70 | false | true (0.1) | str_pdb70 |
     | str_pdb | string (uniref90) | default | default | pdb | false | true (0.1) | str_pdb |
     | str_comp | string (uniref90) | default | default | pdb_complex | false | true (0.1) | str_comp |
     | str_inter_prot_sto | string (uniprot) | default | pdb_seqres | false | true (0.1) | str_iter_uniprot_sto |
     | DeepMSA2 predictors | deepmsa | deepmsa | pdb_seqres | false | true (0.1) | DeepMSA2 predictors | 
     | def_drop_s | default | default | pdb_seqres (default) | true | true (0.1) | AFSample | 
     | def_drop_nos | default | default | pdb_seqres (default) | true | false | AFSample | 
     | def_notemp | default | default | no templates | false | true (0.1) | AFSample |
     | def_notemp_drop_s | default | default | no templates | true | true (0.1) | AFSample |
     | def_notemp_drop_nos | default | default | no templates | true | false | AFSample |
     | def_mul_esm_msa | default + esm MSA | default | pdb_seqres (default) | false | true (0.1) | def_mul_esm_msa |
     | folds_iter (pdb + af_db) | Foldseek-found paired alignments (pdb + af_db) | input msa + Foldseek-found paired alignments | Foldseek-found templates | false | true (0.1) | Structural-alignment based methods |
     | folds_iter_nop (pdb + af_db) | None | input msa + Foldseek-found paired alignments | Foldseek-found templates | false | true (0.1) | Structural-alignment based methods |
     | folds_iter_not (pdb + af_db) | Foldseek-found paired alignments| input msa + Foldseek-found paired alignments | no template | false | true (0.1) | Structural-alignment based methods |
     | folds_iter_esm (esm_atlas) | Foldseek-found paired alignments | input msa + Foldseek-found paired alignments | pdb_seqres (default) | false | true (0.1) | Structural-alignment based methods |
     | folds_iter_esm_nop (esm_atlas) | None | input msa + Foldseek-found paired alignments | pdb_seqres (default) | false | true (0.1) | Structural-alignment based methods |
     | folds_iter_esm_not (esm_atlas) | Foldseek-found paired alignments | input msa + Foldseek-found paired alignments | no template | false | true (0.1) | Structural-alignment based methods |
     | AFProfile | default (uniprot) | default | pdb_seqres | false | true (0.1) | AFProfile using the feautres from default_multimer | 

- AlphaFold-Multimer-based predictors (homo-multimer)

     | Predictor  | $MSA_{paired}$ | $MSA_{unpaired}$ | Template | dropout | dropout in structure module | Description |
     | -------------| ---------------| ------ | ------ | ---------------| ------ | ------ |
     | default_multimer | default (uniprot) | default | pdb_seqres | false | true (0.1) | default AlphaFold-Multimer |
     | def_mul_nopair | None | default | pdb_seqres | false | true (0.1) | default_multimer no msa pairing | 
     | def_mul_struct | default (uniprot) | default | structural template (foldseek) |  false | true (0.1) | def_mul_struct |
     | def_mul_tmsearch | default (uniprot) | default | structural template (tmsearch) |  false | true (0.1) | def_mul_tmsearch |
     | def_mul_pdb70 | default (uniprot) | default | pdb70 | false | true (0.1) | def_mul_pdb70 |
     | def_mul_pdb | default (uniprot) | default | pdb | false | true (0.1) | def_mul_pdb |
     | def_mul_comp | default (uniprot) | default | pdb_complex | false | true (0.1) | def_mul_comp |
     | uniclust_ox_a3m | species (unclust30) | default | pdb_seqres | false | true (0.1) | uniclust_oxmatch_a3m |
     | pdb_iter_ref_a3m | pdb (uniref30) | default | pdb_seqres | false | true (0.1) | pdb_iter_uniref_a3m |
     | pdb_iter_ref_sto | pdb (uniref90) | default | pdb_seqres | false | true (0.1) | pdb_iter_uniref_sto | 
     | pdb_iter_prot_sto | pdb (uniprot) | default | pdb_seqres | false | true (0.1) | pdb_iter_uniprot_sto |
     | spec_iter_ref_a3m | species (uniref30) | default | pdb_seqres | false | true (0.1) | spec_iter_uniref_a3m |
     | spec_struct | species (uniref30) | default | structural template (foldseek) | false | true (0.1) | spec_struct |
     | spec_pdb70 | species (uniref30) | default | pdb70 | false | true (0.1) | spec_pdb70 |
     | spec_pdb | species (uniref30) | default | pdb | false | true (0.1) | spec_pdb |
     | spec_comp | species (uniref30) | default | pdb_complex | false | true (0.1) | spec_comp |
     | spec_iter_ref_sto | species (uniref90) | default | pdb_seqres | false | true (0.1) | spec_iter_uniref_sto |
     | spec_iter_prot_sto | species (uniprot) | default | pdb_seqres | false | true (0.1) | spec_iter_uniprot_sto |
     | DeepMSA2 predictors | deepmsa | deepmsa | pdb_seqres | false | true (0.1) | DeepMSA2 predictors | 
     | def_drop_s | default | default | pdb_seqres (default) | true | true (0.1) | AFSample | 
     | def_drop_nos | default | default | pdb_seqres (default) | true | false | AFSample | 
     | def_notemp | default | default | no templates | false | true (0.1) | AFSample |
     | def_notemp_drop_s | default | default | no templates | true | true (0.1) | AFSample |
     | def_notemp_drop_nos | default | default | no templates | true | false | AFSample |
     | folds_iter (pdb + af_db) | Foldseek-found paired alignments (pdb + af_db) | input msa + Foldseek-found paired alignments | Foldseek-found templates | false | true (0.1) | Structural-alignment based methods |
     | folds_iter_not (pdb + af_db) | Foldseek-found paired alignments| input msa + Foldseek-found paired alignments | no template | false | true (0.1) | Structural-alignment based methods |
     | folds_iter_o (pdb + af_db) | Foldseek-found paired alignments (pdb + af_db) | input msa + Foldseek-found paired alignments | Foldseek-found templates | false | true (0.1) | Structural-alignment based methods |
     | folds_iter_o_not (pdb + af_db) | Foldseek-found paired alignments| input msa + Foldseek-found paired alignments | no template | false | true (0.1) | Structural-alignment based methods |
     | folds_iter_esm (esm_atlas) | Foldseek-found paired alignments | input msa + Foldseek-found paired alignments | pdb_seqres (default) | false | true (0.1) | Structural-alignment based methods |
     | folds_iter_esm_not (esm_atlas) | Foldseek-found paired alignments | input msa + Foldseek-found paired alignments | no template | false | true (0.1) | Structural-alignment based methods |
     | folds_iter_o_esm (esm_atlas) | Foldseek-found paired alignments | input msa + Foldseek-found paired alignments | pdb_seqres (default) | false | true (0.1) | Structural-alignment based methods |
     | folds_iter_esm_o_not (esm_atlas) | Foldseek-found paired alignments | input msa + Foldseek-found paired alignments | no template | false | true (0.1) | Structural-alignment based methods |
     | AFProfile | default (uniprot) | default | pdb_seqres | false | true (0.1) | AFProfile using the feautres from default_multimer | 

- Post-AlphaFold2-based predictors (requires models generated by default as inputs)

     | Predictor  | MSA | Template | dropout | dropout in structure module | Description |
     | -------- | ----- | ----- | ----- | ----- | ----- | ----- |
     | def_mul_refine | Default + Foldseek (pdb + af_db) | Foldseek | false | true (0.1) | Foldseek refinement using the top-ranked models generated by default alphafold2, number of input models: 5; number of output models: 5 |

Number of models to generate:

About 500 models per AlphaFold-Multimer-based variant (MSA, template, dropout), select the top 100 models for relaxation and further ranking.

For hard targets, use AFsample to generate 1000 models per variant (total 5000) in AFsample, select the top 200 for relaxation per variant and further ranking. For other predictors, increase to 1000 models per AlphaFold2-based variants, and select the top 200 models for relaxation.

- Non-AlphaFold-Multimer-based predictors

| Predictor  | Note |
| -------------| -----| 
| ESMFold | ESMFold|


### Hellbender
     ```
     cd /cluster/pixstor/chengji-lab/bml_casp16/MULTICOM4
     source /cluster/pixstor/chengji-lab/bml_casp16/anaconda3/bin/activate
     conda activate multicom4 
     export PYTHONPATH=/cluster/pixstor/chengji-lab/bml_casp16/MULTICOM4
     export PERL5LIB=/cluster/pixstor/chengji-lab/bml_casp16/tools/DNCON4/tools/SCRATCH-1D_1.1
     
     # Need at least 300G memory, can only be run on the H100 nodes
     python bin/multimer/multimer_subunit_alignment.py --option_file bin/db_option_default_hell --fasta_path examples/fasta/H1106.fasta --output_dir examples/H1106

     # Run the alphafold2 predictors, the jobs will be automatically submitted
     # Can run multiple times to check whether some of the predictors failed to generate models
     python bin/multimer/multimer_subunit_af.py --option_file bin/db_option_default_hell --fasta_path examples/fasta/H1106.fasta --output_dir examples/H1106
     
     # After the default alphafold2 models have been generated, run the script again to run the post-af2 predictors for each subunits
     python bin/multimer/multimer_subunit_af.py --option_file bin/db_option_default_hell --fasta_path examples/fasta/H1106.fasta --output_dir examples/H1106

     # Generate multimer alignments and templates
     # Run default_multimer first, since other methods require some files from default_multimer
     # Didn't submit any jobs for subunits
     python bin/multimer/multimer_alignment_default.py --option_file bin/db_option_default_hell --fasta_path examples/fasta/H1106.fasta --output_dir examples/H1106/

     python bin/multimer/heteromer_af.py --option_file bin/db_option_default_hell --fasta_path examples/fasta/H1106.fasta --output_dir examples/H1106/
     ```


